### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by itshaydendev**
> This PR is still a WIP, so some details of it may change over time as needed

## Change information

- NodeJS version requirement increased to LTS at the time of writing (v12.16.1)
- Added `prettier` for code formatting. Can be paired up with ESLint
- Updated Docker Compose configuration to v3 (was on v2 still)
- Replaced tslint (deprecated) with typescript-eslint
- Prefer ormconfig over connection file

### CODE REVIEW

**Comment by itshaydendev** [comment]
Perhaps we could ditch the configuration in `config/connection.ts` in favour of something like [an ormconfig file](https://github.com/typeorm/typeorm/blob/master/docs/using-ormconfig.md)? It would let us leave database choice to the user instead of specifying MongoDB (or at least making it easier for them). TypeORM is powerful like that.

**Comment by itshaydendev** [comment]
Also, topic of TypeORM - Perhaps we could transition to Active Record over using getRepository so much?

**Comment by unix** [comment]
ACK, I voted for these:

- [x] Require LTS
- [x] `prettier` support
- [x]  update Docker Compose 
- [x] `ormconfig` file is good solution

And, file `ormconfig.ext` maybe require multiple envs. (I've invited you into this project. Pay attention to the email)

File: docker-compose.yml [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
     image: tutum/mongodb:3.0
     ports:
-    - "27017:27017"
-    - "28017:28017"
+      - '27017:27017'
+      - '28017:28017'
     restart: always
     volumes:
-      - koa_ex_db:/data/db
+      - db:/data/db
```
Comment (unix) [inline docker-compose.yml]: If I had many `volumes` or `containers`, is it difficult to distinguish them?

Comment (itshaydendev) [reply docker-compose.yml]: No, it prefixes containers with `foldername_` unless you specify `containerName` on each definition

Comment (itshaydendev) [reply docker-compose.yml]: For example, in a directory called `myapp`, all the containers will be like `myapp_db_1`, etc.

**Comment by unix** [comment]
> Voted? Like, you wanted those?

Yes. I think these changes are valuable.

File: .gitignore [hunk 2]
```diff
@@ -57,3 +57,5 @@ typings/
 # dotenv environment variables file
 .env
 
+# ormconfig file
+ormconfig.json
```
Comment (unix) [inline .gitignore]: I think this file can be added to version control, In the file, we can write like this:

```js
dbname: Environment.DB.name,    // the value from process.env.NODE_ENV or other envs
```

This file does not contain security information. Of course, you're still at work. It's just a suggestion.

File: app/entities/sessions.entity.ts [hunk 3]
```diff
@@ -1,6 +1,10 @@
 import { MinLength, IsNotEmpty } from 'class-validator'
 import {
-  Entity, BaseEntity, ObjectIdColumn, CreateDateColumn, UpdateDateColumn,
+  Entity,
+  BaseEntity,
+  ObjectIdColumn,
+  CreateDateColumn,
+  UpdateDateColumn
```
Comment (unix) [inline app/entities/sessions.entity.ts]:
```suggestion
  UpdateDateColumn,
```
The last line always has a comma, keep up with the old style.

File: configs/connection.ts [hunk 4]
```diff
@@ -1,26 +1,11 @@
-import * as entities from 'entities'
 import * as bootstrap from './bootstrap'
 import { print } from './utils'
-import { Environment } from './environments'
 import { createConnection } from 'typeorm'
-const mongo = Environment.mongo
 
+const mongoConnect = createConnection().then(() =>
```
Comment (unix) [inline configs/connection.ts]: This function should be rename? (Now we can use multiple databases)

File: .gitignore (hunk 2 above)
Comment (itshaydendev) [reply .gitignore]: Yes it does, if you look, there's an `ormconfig.example.json` file which can be copied and edited. It's a JSON file, not a JS file. And it's bad practice to store secrets in code anyway

File: configs/environments/index.ts [hunk 5]
```diff
@@ -5,13 +5,12 @@ const parsedEnvs = bootstrap.before()
 const isProd = process.env.NODE_ENV === 'production'
 const env = isProd ? production : development
 
-
 // Preference of environmental variables.
-Object.keys(env.mongo).forEach(key => {
+Object.keys(env.mongo).forEach((key) => {
   env.mongo[key] = process.env[key] || env.mongo[key]
```
Comment (unix) [inline configs/environments/index.ts]: Maybe we can cover all environment variables, not just `mongo`, or remove these lines of code? (we don't seem to need to declare `mongo` alone)

If we declare environment variables in file `.env`, these variables are assigned to `process.env`. this is expected behavior:

<br />

In the `.env` file
```
//  .env
DB_NAME: '123'
```
And in the `config/environments/production.ts`
```
DB_NAME: '000'
```

<br />

`process.env` at runtime:
```
console.log(process.env.DB_NAME)  // => 123
```

<br />

When used in other files:
```
import { DB_NAME } from './environments'

console.log(Environment.SECRET_NAME) // => 123
```

File: app/entities/sessions.entity.ts (hunk 3 above)
Comment (itshaydendev) [reply app/entities/sessions.entity.ts]: Alright, I'll put trailing commas into the Prettier config

File: configs/connection.ts (hunk 4 above)
Comment (itshaydendev) [reply configs/connection.ts]: True, true

File: .gitignore (hunk 2 above)
Comment (unix) [reply .gitignore]: All secrets related to production environment are in `.env` file, other secrets of development and debugging are in file `configs/environments`.

Is it possible for us to do so?

<br />

In `.env` file:
```
MY_SECRET: 'hi',
```

<br />

In `ormconfig.js` file:
```
import { env } from 'config/environemnts'

DB_SECRET: env.MY_SECRET,
```

Comment (unix) [reply .gitignore]: Because different databases need to be automatically linked when starting applications in multiple environments.

**Comment by leekoho** [comment]
> Perhaps we could ditch the configuration in `config/connection.ts` in favour of something like [an ormconfig file](https://github.com/typeorm/typeorm/blob/master/docs/using-ormconfig.md)? It would let us leave database choice to the user instead of specifying MongoDB (or at least making it easier for them). TypeORM is powerful like that.

atm, typeorm for mongodb is poor

**Comment by unix** [comment]
Hi, @itshaydendev , there are a few solutions to discuss:

**1. Where to configure databases of multiple environments?**

  I suggest adding  `ormconfig.[ext]` to version control, the content of file `ormconfig.[ext]` does not contain any private information, as follows:

```js
const env = require('configs/envs')
module.exports = {
  "type": process.env.something,       // use env variables directly
  "host": env.host,                                 // or use environment file
}
```

  All development and test configuration information is written in file `configs/envs/*.ts`, all information about the prod env is written in file `variables.env` or `[anything].env`. In the prod env, the contents of file `.env` overwrite the environment variables.

  It's a basic solution, or  do you have a better one?

**2. The readme file needs to be updated to guide developers how to set configuration information.**

Or do I initiate the work related to `ormconfig`? You decide what to do.

**Comment by unix** [comment]
Hi, @itshaydendev any update here?

**Comment by itshaydendev** [comment]
Hey, sorry, been a bit bogged down with work.

I think what I'm gonna do is (because I've not for a while) re-look over the code here and then I'll get back to you.

Sorry again

**Comment by unix** [comment]
@itshaydendev 
Thank you for your contribution. Because this PR is not active, I close it.

And update for anyone:
  - Version `3.0` is released. I redesigned the database connection, now it's easier to use the database.

**Pull request closed without merging**